### ISSUE TITLE
[luci/pass] FuseInstanceNormPass with 3D causes shape inference failure

### ISSUE THREAD

**Issue by seanshpark** [Samsung/ONE#14386]
`FuseInstanceNormPass` with 3D (as GroupNorm from ONNX) causes shape inference failure
- related #13778

> circle2circle: ERROR: Internal Exception. Cannot produce expand_dimension of two shapes [/home/jenkins/jenkins_agent/workspace/nnfw/master/daily-publish-package/code/compiler/luci/service/src/CircleShapeInferenceHelper.cpp:90]

**Comment by seanshpark** [comment]
Our internal model has shared `const_as_beta` that should not update in-place but shape is change from `[1,32,1]` to `[32]`.
- this model looks like it's not trained so all the `beta` have 0.0f and shared nodes

**Comment by seanshpark** [comment]
With the fix,

> record-minmax: ERROR: luci-interpreter/src/kernels/InstanceNorm.cpp:56(gamma()->shape().dim(0) == input()->shape().dim(1) || gamma()->shape().dim(0) == 1) was not true.

**Comment by seanshpark** [comment]
InstanceNorm after fusion

![image](https://github.com/user-attachments/assets/2c0168c6-c48a-4dd0-9039-5848052d492b)

need to check gamma, beta shape is OK...

**Comment by seanshpark** [comment]
For 4D `InstanceNorm`, as from `NHWC`, last dim is channel
For 3D `InstanceNorm`, source ONNX uses channel as dim(1) and this doesn't change when converted to tflite/circle.
So we need to use `dim(1)` as channel for this case.

**Comment by mhs4670go** [comment]
> So we need to use dim(1) as channel for this case.

Actually, I thought this and below resolves the error. Therefore, just curiosity, https://github.com/Samsung/ONE/pull/14387/ is necessary?

```diff
@@ -1180,12 +1195,16 @@ bool is_add_input_mul_sub3d(luci::CircleAdd *add)
   if (!luci::fill(&p_mul, &p_sub).with_commutative_args_of(add))
     return false;
 
-  auto sub = dynamic_cast<luci::CircleSub *>(add->y());
+  auto sub = dynamic_cast<luci::CircleSub *>(p_sub);
   if (sub == nullptr)
     return false;
 
-  auto const_as_beta = dynamic_cast<luci::CircleConst *>(sub->x());
-  if (const_as_beta == nullptr || const_as_beta->rank() != 3)
+  p_mul = nullptr;
+  luci::CircleConst *const_as_beta = nullptr;
+  if (!luci::fill(&p_mul, &const_as_beta).with_commutative_args_of(sub))
+    return false;
+
+  if (const_as_beta->rank() != 3)
     return false;
 
   return true;
```

**Comment by seanshpark** [comment]
> Therefore, just curiosity, #14387 is necessary?

Yes I think so, `const_as_beta` itself is shared constant that is used for multiple `Sub` input,

**Comment by mhs4670go** [comment]
> const_as_beta itself is shared constant that is used for multiple Sub input,

I got it. Seems that it changes from modifying the nodes to cloning and modifying the modes as what we always does.

Then, I'm not sure but IMHO, the clone of beta should be done in `reshape_gamma_beta` where nodes are modified because `InstanceNormPattern::match` seems to do just matching dedicated pattern.

**Comment by seanshpark** [comment]
> Then, I'm not sure but IMHO, the clone of beta should be done in reshape_gamma_beta ...

Yes, the original code was built like this.
Why it's working is, I think, (1) there wasn't many InstNorm nodes in our models, (2) all models was trained and thus had different values in Const nodes so that there was no problem.

We may have to clone the `gamma` or `beta` for all other cases some day...

**Comment by jinevening** [comment]
@seanshpark  BTW, value check was done for instancenorm 3d?

I saw the pattern, and found that it's different from the original semantics (this can be easily found with pen and paper).

![image](https://github.com/user-attachments/assets/d661acce-0ba2-4e1e-954f-4a587ea478a9)

In addition, the problematic tensor is the first input of a Sub Op, whose values are all zeros. Why is it translated to a beta? I think the overall code of instance norm 3d fusion has to be revisited.

**Comment by seanshpark** [comment]
> value check was done for instancenorm 3d?

Yes, test model is `Net_InstanceNorm_008` and value check is done.

> Why is it translated to a beta?

I think the it was from our test ONNX model that @ragmani made.. maybe from ViT ???

> I think the overall code of instance norm 3d fusion has to be revisited.

Um? which part?

**Comment by seanshpark** [comment]
`Net_InstanceNorm_008` with random value fails -_-;

**Comment by jinevening** [comment]
I guess that `Net_InstanceNorm_008` was from the circle model generated by onnx2circle. We may need to do value check between `Net_InstanceNorm_008` and the original onnx model first.

**Comment by seanshpark** [comment]
> Why is it translated to a beta?

In `FuseInstanceNormPass`, this pattern for 3D is version 6, identical as version 5, which has
```
 *         |   const_as_beta   |
 *         |         |         V
 *         |         +------> sub
 *         V                   |
```
I guess, this is what onnx2tf -> tf2tflite made.

**Comment by seanshpark** [comment]
> and found that it's different from the original semantics

Can you please point out what part is wrong? we've to revise `version_5` too.

**Comment by seanshpark** [comment]
> ... circle model generated by onnx2circle.

`onnx2circle` converts to `CircleInstanceNorm` so the exploded graph isn't generated.

**Comment by jinevening** [comment]
Hmm.. It seems that the axis of InstanceNorm with rank 3 is 1.

TF's default axis is -1 (the last dimension). We also followed the convention for InstanceNorm with rank 4, which is aligned with `channel_last` layout that circle uses.
(ref: https://www.tensorflow.org/api_docs/python/tf/keras/layers/GroupNormalization)

Only InstanceNorm with rank3 uses a different layout (`channel_first`). This difference is confusing, and in fact can degrade performance of our backend which works in a `channel_last` manner.

Is it necessary to define rank3 instnorm with axis = 1? If so, what is it used for?

**Comment by seanshpark** [comment]
> Is it necessary to define rank3 instnorm with axis = 1? 

I don't know ... -_-;

> If so, what is it used for?

It's from our customer ...

**Comment by ragmani** [comment]
> Only InstanceNorm with rank3 uses a different layout (channel_first). This difference is confusing, and does not align well with our backend which works in a channel_last manner.

I didn't catch what problem was exactly, but I think it's just a default value in TF. We need to just convert onnx models to circle models . According to https://onnx.ai/onnx/operators/onnx__InstanceNormalization.html#inputs, the dimensions are in the form of (N x C x D1 x D2 … Dn) for non image case. Is it necessary to transpose InstanceNorm with rank 4 to (N x D1 x C)?

**Comment by seanshpark** [comment]
> My concern is that it is difficult to generate code for our backend for rank3 instnorm with axis = 1.

Then lets' discuss this in our internal issue.

**Comment by seanshpark** [comment]
Our internal model fails to compile only with #14388.

**Comment by seanshpark** [comment]
close for now. add new issue when backend decides it's policy.
